{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/bulk-uranus/gwoelflein/toolmaker/ToolArena/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[ToolDefinition(name='cobra_extract_features', repo=Repository(name='COBRA', url='https://github.com/KatherLab/COBRA', branch=None, commit=None, env=[EnvironmentVariable(name='HF_TOKEN', value='${env:HF_TOKEN}')]), requires='cuda', papers=['lenz2025cobra'], category='pathology', description='Perform feature extraction on an input image using COBRA. The provided tile features have 1.0 mpp (224 microns / 224 px per patch).', arguments=[ArgumentDefinition(name='output_dir', description='Path to the output folder where the features will be saved', type='str'), ArgumentDefinition(name='input_dir', description='Path to the input folder containing the tile features', type='str')], returns=[ArgumentDefinition(name='num_processed_slides', description='The number of slides that were processed', type='int')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_features-brca-single'), ArgumentValue(name='input_dir', value='/mount/input/TCGA-BRCA-Virchow2-features-10x/')], mount=[Mount(source='TCGA-BRCA-Virchow2-features-10x/TCGA-BH-A0HQ-01Z-00-DX1.0921FCEF-20A2-4D4B-A198-91AF9F6C814C.h5', target='TCGA-BRCA-Virchow2-features-10x/TCGA-BH-A0HQ-01Z-00-DX1.0921FCEF-20A2-4D4B-A198-91AF9F6C814C.h5')]), test_invocations=[ToolInvocation(name='crc', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_features-crc'), ArgumentValue(name='input_dir', value='/mount/input/TCGA-CRC-Virchow2-features-10x')], mount=[Mount(source='TCGA-CRC-Virchow2-features-10x', target='TCGA-CRC-Virchow2-features-10x')]), ToolInvocation(name='brca', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_features-brca'), ArgumentValue(name='input_dir', value='/mount/input/TCGA-BRCA-Virchow2-features-10x')], mount=[Mount(source='TCGA-BRCA-Virchow2-features-10x', target='TCGA-BRCA-Virchow2-features-10x')]), ToolInvocation(name='brca-single', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_features-brca-single'), ArgumentValue(name='input_dir', value='/mount/input/TCGA-BRCA-Virchow2-features-10x/')], mount=[Mount(source='TCGA-BRCA-Virchow2-features-10x/TCGA-BH-A0HQ-01Z-00-DX1.0921FCEF-20A2-4D4B-A198-91AF9F6C814C.h5', target='TCGA-BRCA-Virchow2-features-10x/TCGA-BH-A0HQ-01Z-00-DX1.0921FCEF-20A2-4D4B-A198-91AF9F6C814C.h5')])], note=None),\n",
       " ToolDefinition(name='cobra_heatmaps', repo=Repository(name='COBRA', url='https://github.com/KatherLab/COBRA', branch=None, commit=None, env=[EnvironmentVariable(name='HF_TOKEN', value='${env:HF_TOKEN}')]), requires='cuda', papers=['lenz2025cobra'], category='pathology', description='Create unsupervised heatmaps using COBRA. The provided tile features have 1.0 mpp (224 microns / 224 px per patch).', arguments=[ArgumentDefinition(name='output_dir', description='Path to the output folder where the features will be saved', type='str'), ArgumentDefinition(name='slide_dir', description='Path to the input folder containing the WSIs', type='str'), ArgumentDefinition(name='tile_features_dir', description='Path to the input folder containing the tile features', type='str')], returns=[ArgumentDefinition(name='num_heatmaps', description='The number of heatmaps that were generated', type='int'), ArgumentDefinition(name='byte_size_heatmaps', description='The number of total bytes of the generated heatmaps', type='int')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_heatmaps-crc'), ArgumentValue(name='tile_features_dir', value='/mount/input/TCGA-CRC-Virchow2-features-10x'), ArgumentValue(name='slide_dir', value='/mount/input/crc-wsi')], mount=[Mount(source='TCGA-CRC-Virchow2-features-10x', target='TCGA-CRC-Virchow2-features-10x'), Mount(source='crc-wsi', target='crc-wsi')]), test_invocations=[ToolInvocation(name='crc', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_heatmaps-crc'), ArgumentValue(name='tile_features_dir', value='/mount/input/TCGA-CRC-Virchow2-features-10x'), ArgumentValue(name='slide_dir', value='/mount/input/crc-wsi')], mount=[Mount(source='TCGA-CRC-Virchow2-features-10x', target='TCGA-CRC-Virchow2-features-10x'), Mount(source='crc-wsi', target='crc-wsi')]), ToolInvocation(name='brca', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_heatmaps-brca'), ArgumentValue(name='tile_features_dir', value='/mount/input/TCGA-BRCA-Virchow2-features-10x'), ArgumentValue(name='slide_dir', value='/mount/input/brca-wsi')], mount=[Mount(source='TCGA-BRCA-Virchow2-features-10x', target='TCGA-BRCA-Virchow2-features-10x'), Mount(source='brca-wsi', target='brca-wsi')]), ToolInvocation(name='brca-single', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_heatmaps-brca-single'), ArgumentValue(name='tile_features_dir', value='/mount/input/TCGA-BRCA-Virchow2-features-10x'), ArgumentValue(name='slide_dir', value='/mount/input/brca-wsi')], mount=[Mount(source='TCGA-BRCA-Virchow2-features-10x/TCGA-BH-A0HQ-01Z-00-DX1.0921FCEF-20A2-4D4B-A198-91AF9F6C814C.h5', target='TCGA-BRCA-Virchow2-features-10x/TCGA-BH-A0HQ-01Z-00-DX1.0921FCEF-20A2-4D4B-A198-91AF9F6C814C.h5'), Mount(source='brca-wsi/TCGA-BH-A0HQ-01Z-00-DX1.0921FCEF-20A2-4D4B-A198-91AF9F6C814C.svs', target='brca-wsi/TCGA-BH-A0HQ-01Z-00-DX1.0921FCEF-20A2-4D4B-A198-91AF9F6C814C.svs')])], note=None),\n",
       " ToolDefinition(name='conch_extract_features', repo=Repository(name='CONCH', url='https://github.com/mahmoodlab/CONCH', branch=None, commit='171f2be', env=[EnvironmentVariable(name='HF_TOKEN', value='${env:HF_TOKEN}')]), requires='cuda', papers=['lu2024conch'], category='pathology', description='Perform feature extraction on an input image using CONCH.', arguments=[ArgumentDefinition(name='input_image', description='Path to the input image', type='str')], returns=[ArgumentDefinition(name='features', description='The feature vector extracted from the input image, as a list of floats', type='list')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='input_image', value='/mount/input/TUM-TCGA-ACRLPPQE.tif')], mount=[Mount(source='TUM-TCGA-ACRLPPQE.tif', target='TUM-TCGA-ACRLPPQE.tif')]), test_invocations=[ToolInvocation(name='tif', arguments=[ArgumentValue(name='input_image', value='/mount/input/MUC/MUC-TCGA-ACCPKIPN.tif')], mount=[Mount(source='MUC-TCGA-ACCPKIPN.tif', target='MUC/MUC-TCGA-ACCPKIPN.tif')]), ToolInvocation(name='png', arguments=[ArgumentValue(name='input_image', value='/mount/input/TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.png')], mount=[Mount(source='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.png', target='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.png')]), ToolInvocation(name='jpg', arguments=[ArgumentValue(name='input_image', value='/mount/input/TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.jpg')], mount=[Mount(source='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.jpg', target='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.jpg')])], note=None),\n",
       " ToolDefinition(name='cytopus_db', repo=Repository(name='Cytopus', url='https://github.com/wallet-maker/cytopus', branch=None, commit='638dd91', env=[]), requires='cpu', papers=['kunes2023cytopus'], category='genomics_proteomics', description='Initialize the Cytopus KnowledgeBase and generate a JSON file containing a nested dictionary with gene set annotations organized by cell type, suitable for input into the Spectra library.', arguments=[ArgumentDefinition(name='celltype_of_interest', description='List of cell types for which to retrieve gene sets', type='list'), ArgumentDefinition(name='global_celltypes', description='List of global cell types to include in the JSON file.', type='list'), ArgumentDefinition(name='output_file', description='Path to the file where the output JSON file should be stored.', type='str')], returns=[ArgumentDefinition(name='keys', description='The list of keys in the produced JSON file.', type='list')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='celltype_of_interest', value=['B_memory', 'B_naive', 'CD4_T', 'CD8_T', 'DC', 'ILC3', 'MDC', 'NK', 'Treg', 'gdT', 'mast', 'pDC', 'plasma']), ArgumentValue(name='global_celltypes', value=['all-cells', 'leukocyte']), ArgumentValue(name='output_file', value='/mount/output/Spectra_dict.json')], mount=[]), test_invocations=[ToolInvocation(name='B_and_CD4_T', arguments=[ArgumentValue(name='celltype_of_interest', value=['B', 'CD4_T']), ArgumentValue(name='global_celltypes', value=['all-cells', 'leukocyte']), ArgumentValue(name='output_file', value='/mount/output/Spectra_dict.json')], mount=[]), ToolInvocation(name='leukocytes', arguments=[ArgumentValue(name='celltype_of_interest', value=['B_memory', 'B_naive', 'CD4_T', 'CD8_T', 'DC', 'ILC3', 'MDC', 'NK', 'Treg', 'gdT', 'mast', 'pDC', 'plasma']), ArgumentValue(name='global_celltypes', value=['leukocyte']), ArgumentValue(name='output_file', value='/mount/output/Spectra_dict.json')], mount=[]), ToolInvocation(name='Treg_and_plasma_and_B_naive', arguments=[ArgumentValue(name='celltype_of_interest', value=['Treg', 'plasma', 'B_naive']), ArgumentValue(name='global_celltypes', value=['leukocyte']), ArgumentValue(name='output_file', value='/mount/output/Spectra_dict.json')], mount=[])], note='The information on how to do this is in: https://github.com/wallet-maker/cytopus/blob/main/notebooks/KnowledgeBase_queries_colaboratory.ipynb\\n'),\n",
       " ToolDefinition(name='eagle_extract_features', repo=Repository(name='EAGLE', url='https://github.com/KatherLab/EAGLE', branch='simple_feature_extraction', commit=None, env=[]), requires='cuda', papers=['neidlinger2025eagle'], category='pathology', description='Perform slide-level feature extraction using EAGLE, given tile-level features.', arguments=[ArgumentDefinition(name='output_dir', description='Path to the output folder where the features will be saved', type='str'), ArgumentDefinition(name='feature_dir_weighting', description='Path to the input folder containing the tile features to create the weighting (chief-CTP). Files are in *.h5 format, one file per slide.', type='str'), ArgumentDefinition(name='feature_dir_aggregation', description='Path to the input folder containing the tile features that will be aggregated (Virchow2). Files are in *.h5 format, one file per slide.', type='str')], returns=[ArgumentDefinition(name='num_processed_slides', description='The number of slides that were processed', type='int')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_features-crc'), ArgumentValue(name='feature_dir_aggregation', value='/mount/input/TCGA-CRC-Virchow2-features'), ArgumentValue(name='feature_dir_weighting', value='/mount/input/TCGA-CRC-ChiefCTP-features')], mount=[Mount(source='TCGA-CRC-Virchow2-features', target='TCGA-CRC-Virchow2-features'), Mount(source='TCGA-CRC-ChiefCTP-features', target='TCGA-CRC-ChiefCTP-features')]), test_invocations=[ToolInvocation(name='crc', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_features-crc'), ArgumentValue(name='feature_dir_aggregation', value='/mount/input/TCGA-CRC-Virchow2-features'), ArgumentValue(name='feature_dir_weighting', value='/mount/input/TCGA-CRC-ChiefCTP-features')], mount=[Mount(source='TCGA-CRC-Virchow2-features', target='TCGA-CRC-Virchow2-features'), Mount(source='TCGA-CRC-ChiefCTP-features', target='TCGA-CRC-ChiefCTP-features')]), ToolInvocation(name='brca', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_features-brca'), ArgumentValue(name='feature_dir_aggregation', value='/mount/input/TCGA-BRCA-Virchow2-features'), ArgumentValue(name='feature_dir_weighting', value='/mount/input/TCGA-BRCA-ChiefCTP-features')], mount=[Mount(source='TCGA-BRCA-Virchow2-features', target='TCGA-BRCA-Virchow2-features'), Mount(source='TCGA-BRCA-ChiefCTP-features', target='TCGA-BRCA-ChiefCTP-features')]), ToolInvocation(name='brca-single', arguments=[ArgumentValue(name='output_dir', value='/mount/output/cobra_features-brca-single'), ArgumentValue(name='feature_dir_aggregation', value='/mount/input/TCGA-BRCA-Virchow2-features'), ArgumentValue(name='feature_dir_weighting', value='/mount/input/TCGA-BRCA-ChiefCTP-features')], mount=[Mount(source='TCGA-BRCA-Virchow2-features/TCGA-3C-AALJ-01Z-00-DX2.62DFE56B-B84C-40F9-9625-FCB55767B70D.h5', target='TCGA-BRCA-Virchow2-features/TCGA-3C-AALJ-01Z-00-DX2.62DFE56B-B84C-40F9-9625-FCB55767B70D.h5'), Mount(source='TCGA-BRCA-ChiefCTP-features/TCGA-3C-AALJ-01Z-00-DX2.62DFE56B-B84C-40F9-9625-FCB55767B70D.h5', target='TCGA-BRCA-ChiefCTP-features/TCGA-3C-AALJ-01Z-00-DX2.62DFE56B-B84C-40F9-9625-FCB55767B70D.h5')])], note=None),\n",
       " ToolDefinition(name='esm_fold_predict', repo=Repository(name='ESM', url='https://github.com/facebookresearch/esm', branch=None, commit='2b36991', env=[]), requires='cuda', papers=['verkuil2022esm1', 'hie2022esm2'], category='genomics_proteomics', description=\"Generate the representation of a protein sequence and the contact map using Facebook Research's pretrained esm2_t33_650M_UR50D model.\", arguments=[ArgumentDefinition(name='sequence', description='Protein sequence to for which to generate representation and contact map.', type='str')], returns=[ArgumentDefinition(name='sequence_representation', description='Token representations for the protein sequence as a list of floats, i.e. a 1D array of shape L where L is the number of tokens.', type='list'), ArgumentDefinition(name='contact_map', description='Contact map for the protein sequence as a list of list of floats, i.e. a 2D array of shape LxL where L is the number of tokens.', type='list')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='sequence', value='MKTVRQERLKSIVRILERSKEPVSGAQLAEELSVSRQVIVQDIAYLRSLGYNIVATPRGYVLAGG')], mount=[]), test_invocations=[ToolInvocation(name='protein2', arguments=[ArgumentValue(name='sequence', value='KALTARQQEVFDLIRDHISQTGMPPTRAEIAQRLGFRSPNAAEEHLKALARKGVIEIVSGASRGIRLLQEE')], mount=[]), ToolInvocation(name='protein2_with_mask', arguments=[ArgumentValue(name='sequence', value='KALTARQQEVFDLIRD<mask>ISQTGMPPTRAEIAQRLGFRSPNAAEEHLKALARKGVIEIVSGASRGIRLLQEE')], mount=[]), ToolInvocation(name='protein3', arguments=[ArgumentValue(name='sequence', value='K A <mask> I S Q')], mount=[])], note='This repository does not contain any simple CLI functions, but examples.  We ask the model to re-implement one of the examples.'),\n",
       " ToolDefinition(name='flowmap_overfit_scene', repo=Repository(name='FlowMap', url='https://github.com/dcharatan/flowmap', branch=None, commit='578a515', env=[]), requires='cuda', papers=['smith2024flowmap'], category='misc', description='Overfit FlowMap on an input scene to determine camera extrinsics for each frame in the scene.', arguments=[ArgumentDefinition(name='input_scene', description='Path to the directory containing the images of the input scene (just the image files, nothing else)', type='str')], returns=[ArgumentDefinition(name='n', description='The number of images (frames) in the scene', type='int'), ArgumentDefinition(name='camera_extrinsics', description='The camera extrinsics matrix for each of the n frames in the scene, must have a shape of nx4x4 (as a nested python list of floats)', type='list')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='input_scene', value='/mount/input/llff_flower')], mount=[Mount(source='flowmap/llff_flower', target='llff_flower')]), test_invocations=[ToolInvocation(name='llff_fern', arguments=[ArgumentValue(name='input_scene', value='/mount/input/llff_fern')], mount=[Mount(source='flowmap/llff_fern', target='llff_fern')]), ToolInvocation(name='llff_orchids', arguments=[ArgumentValue(name='input_scene', value='/mount/input/llff_orchids')], mount=[Mount(source='flowmap/llff_orchids', target='llff_orchids')])], note=None),\n",
       " ToolDefinition(name='llmaix_extract_text', repo=Repository(name='LLMAIx', url='https://github.com/KatherLab/LLMAIx', branch=None, commit='693564b', env=[]), requires='cpu', papers=['wiest2024llm'], category='llms', description='Preprocess various input files into a standardized structure - a list of the texts. The list can contain multiple texts in case of .csv and .xlsx input files. In some cases, OCR needs to be applied or even enforced to get the correct text from a pdf (or image) document.', arguments=[ArgumentDefinition(name='file_path', description='Path to the input file. Input file can be a .pdf, .csv, .xlsx, .png, .jpg and .docx file.', type='str')], returns=[ArgumentDefinition(name='ocr_text_list', description='The preprocessed document(s), as a list of strings (usually only one string, but in case .csv or .xlsx documents are preprocessed - one per line).', type='list')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='file_path', value='/mount/input/9874563.pdf')], mount=[Mount(source='9874563.pdf', target='9874563.pdf')]), test_invocations=[ToolInvocation(name='pdf_ocr', arguments=[ArgumentValue(name='file_path', value='/mount/input/9874563.pdf')], mount=[Mount(source='9874563.pdf', target='9874563.pdf')]), ToolInvocation(name='pdf_ocr_force', arguments=[ArgumentValue(name='file_path', value='/mount/input/9874562.pdf')], mount=[Mount(source='9874562.pdf', target='9874562.pdf')]), ToolInvocation(name='csv', arguments=[ArgumentValue(name='file_path', value='/mount/input/data.csv')], mount=[Mount(source='data.csv', target='data.csv')]), ToolInvocation(name='xlsx', arguments=[ArgumentValue(name='file_path', value='/mount/input/data.xlsx')], mount=[Mount(source='data.xlsx', target='data.xlsx')]), ToolInvocation(name='image', arguments=[ArgumentValue(name='file_path', value='/mount/input/9874563.png')], mount=[Mount(source='9874563.png', target='9874563.png')])], note=None),\n",
       " ToolDefinition(name='medsam_inference', repo=Repository(name='MedSAM', url='https://github.com/bowang-lab/MedSAM', branch=None, commit='b9db486', env=[]), requires='cuda', papers=['ma2024medsam'], category='radiology', description='Use the trained MedSAM model to segment the given abdomen CT scan.', arguments=[ArgumentDefinition(name='image_file', description='Path to the abdomen CT scan image.', type='str'), ArgumentDefinition(name='bounding_box', description='Bounding box to segment (list of 4 integers).', type='list'), ArgumentDefinition(name='segmentation_file', description='Path to where the segmentation image should be saved.', type='str')], returns=[], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='image_file', value='/mount/input/my_image.jpg'), ArgumentValue(name='bounding_box', value=[25, 100, 155, 155]), ArgumentValue(name='segmentation_file', value='/mount/output/segmented_image.png')], mount=[Mount(source='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.jpg', target='my_image.jpg')]), test_invocations=[ToolInvocation(name='cucumber', arguments=[ArgumentValue(name='image_file', value='/mount/input/cucumber.jpg'), ArgumentValue(name='bounding_box', value=[25, 100, 155, 155]), ArgumentValue(name='segmentation_file', value='/mount/output/segmented_image.png')], mount=[Mount(source='cucumber.jpg', target='cucumber.jpg')]), ToolInvocation(name='other_output_file', arguments=[ArgumentValue(name='image_file', value='/mount/input/cucumber.jpg'), ArgumentValue(name='bounding_box', value=[25, 100, 155, 155]), ArgumentValue(name='segmentation_file', value='/mount/output/some_other_file.png')], mount=[Mount(source='cucumber.jpg', target='cucumber.jpg')]), ToolInvocation(name='png', arguments=[ArgumentValue(name='image_file', value='/mount/input/image2.png'), ArgumentValue(name='bounding_box', value=[25, 100, 155, 155]), ArgumentValue(name='segmentation_file', value='/mount/output/segmented_image.png')], mount=[Mount(source='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.png', target='image2.png')])], note=None),\n",
       " ToolDefinition(name='medsss_generate', repo=Repository(name='MedSSS', url='https://github.com/pixas/MedSSS', branch=None, commit='ebbfd02', env=[EnvironmentVariable(name='HF_TOKEN', value='${env:HF_TOKEN}')]), requires='cuda', papers=['jiang2025medsss'], category='llms', description='Given a user message, generate a response using the MedSSS_Policy model.', arguments=[ArgumentDefinition(name='user_message', description='The user message.', type='str')], returns=[ArgumentDefinition(name='response', description='The response generated by the model.', type='str')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='user_message', value='How to stop a cough?')], mount=[]), test_invocations=[ToolInvocation(name='nsclc', arguments=[ArgumentValue(name='user_message', value='How would you treat a patient with advanced non-small cell lung cancer?')], mount=[]), ToolInvocation(name='motor_vehicle_accident', arguments=[ArgumentValue(name='user_message', value='You are the first responder to a motor vehicle accident. The patient is unconscious and has a suspected spinal injury. What would you do?')], mount=[]), ToolInvocation(name='pediatric_rash', arguments=[ArgumentValue(name='user_message', value='You are a pediatrician seeing a 5 year old with a rash covering their whole chest. What are your first steps?')], mount=[])], note=None),\n",
       " ToolDefinition(name='modernbert_predict_masked', repo=Repository(name='ModernBERT', url='https://github.com/AnswerDotAI/ModernBERT', branch=None, commit='8c57a0f', env=[]), requires='cpu', papers=['warner2024modernbert'], category='llms', description='Given a masked sentence string, predict the original sentence using the pretrained ModernBERT-base model on CPU.', arguments=[ArgumentDefinition(name='input_string', description='The masked sentence string. The masked part is represented by \"[MASK]\"\".', type='str')], returns=[ArgumentDefinition(name='prediction', description='The predicted original sentence (including the predicted masked part)', type='str')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='input_string', value='Paris is the [MASK] of France.')], mount=[]), test_invocations=[ToolInvocation(name='walking', arguments=[ArgumentValue(name='input_string', value='He walked to the [MASK].')], mount=[]), ToolInvocation(name='future_of_ai', arguments=[ArgumentValue(name='input_string', value='The future of AI is [MASK].')], mount=[]), ToolInvocation(name='meaning_of_life', arguments=[ArgumentValue(name='input_string', value='The meaning of life is [MASK].')], mount=[])], note=None),\n",
       " ToolDefinition(name='mopadi_generate_counterfactuals', repo=Repository(name='mopadi', url='https://github.com/KatherLab/mopadi', branch=None, commit='4e76820', env=[EnvironmentVariable(name='HF_TOKEN', value='${env:HF_TOKEN}')]), requires='cuda', papers=['zigutyte2024mopadi'], category='pathology', description=\"Generate counterfactual explanations for the top 3 tiles per patient by manipulating them with a specific amplitude, such that the predicted class of each counterfactual image flips to the opposite class (i.e., the predicted output for the opposite class exceeds 0.9), while avoiding excessive overmanipulation. Use a pretrained diffusion autoencoder according to the cancer type, combined with a corresponding MIL classifier trained to distinguish biologically meaningful histological patterns. You will be provided with the path to the folder containing images, the clinical table with each patient's target label values, and the folder containing pre-extracted features\", arguments=[ArgumentDefinition(name='images_dir', description='Path to the folder containing patient subfolders with image patches', type='str'), ArgumentDefinition(name='feat_path_test', description='Path to the folder containing extracted features for each patient', type='str'), ArgumentDefinition(name='clini_table', description='Path to the XLSX file containing the MSI status of each patient', type='str'), ArgumentDefinition(name='target_label', description='Name of the column in the clinical table that contains classification labels', type='str'), ArgumentDefinition(name='base_dir', description='Path to the output directory where the results will be saved', type='str'), ArgumentDefinition(name='manipulation_levels', description='Amplitude of the manipulation to be applied to the images', type='list'), ArgumentDefinition(name='pretrained_autoenc_name', description='Name of the pretrained diffusion autoencoder model', type='str'), ArgumentDefinition(name='pretrained_clf_name', description='Name of the pretrained classifier', type='str')], returns=[ArgumentDefinition(name='num_counterfactuals', description='The number of counterfactual images that were generated', type='int')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='images_dir', value='/mount/input/images/TCGA-CRC'), ArgumentValue(name='feat_path_test', value='/mount/input/features/TCGA-CRC'), ArgumentValue(name='clini_table', value='/mount/input/TCGA-CRC-DX_CLINI.xlsx'), ArgumentValue(name='target_label', value='isMSIH'), ArgumentValue(name='base_dir', value='/mount/output/counterfactuals_crc_msi'), ArgumentValue(name='manipulation_levels', value=[0.06]), ArgumentValue(name='pretrained_autoenc_name', value='crc_512_model'), ArgumentValue(name='pretrained_clf_name', value='msi')], mount=[Mount(source='images/TCGA-CRC', target='images/TCGA-CRC'), Mount(source='features/TCGA-CRC', target='features/TCGA-CRC'), Mount(source='TCGA-CRC-DX_CLINI.xlsx', target='TCGA-CRC-DX_CLINI.xlsx')]), test_invocations=[ToolInvocation(name='brca_types', arguments=[ArgumentValue(name='images_dir', value='/mount/input/images/TCGA-BRCA'), ArgumentValue(name='feat_path_test', value='/mount/input/features/TCGA-BRCA'), ArgumentValue(name='clini_table', value='/mount/input/TCGA-BRCA-DX_CLINI.csv'), ArgumentValue(name='target_label', value='BRCA_Pathology'), ArgumentValue(name='base_dir', value='/mount/output/counterfactuals_brca_types'), ArgumentValue(name='manipulation_levels', value=[0.06]), ArgumentValue(name='pretrained_autoenc_name', value='brca_512_model'), ArgumentValue(name='pretrained_clf_name', value='type')], mount=[Mount(source='images/TCGA-BRCA', target='images/TCGA-BRCA'), Mount(source='features/TCGA-BRCA', target='features/TCGA-BRCA'), Mount(source='TCGA-BRCA-DX_CLINI.csv', target='TCGA-BRCA-DX_CLINI.csv')]), ToolInvocation(name='liver', arguments=[ArgumentValue(name='images_dir', value='/mount/input/images/Pancancer-liver'), ArgumentValue(name='feat_path_test', value='/mount/input/features/Pancancer'), ArgumentValue(name='clini_table', value='/mount/input/Pancancer_clini.xlsx'), ArgumentValue(name='target_label', value='Type'), ArgumentValue(name='base_dir', value='/mount/output/counterfactuals_liver_types'), ArgumentValue(name='manipulation_levels', value=[0.04]), ArgumentValue(name='pretrained_autoenc_name', value='pancancer_model'), ArgumentValue(name='pretrained_clf_name', value='liver')], mount=[Mount(source='images/Pancancer-liver', target='images/Pancancer-liver'), Mount(source='features/Pancancer', target='features/Pancancer'), Mount(source='Pancancer_clini.xlsx', target='Pancancer_clini.xlsx')]), ToolInvocation(name='lung', arguments=[ArgumentValue(name='images_dir', value='/mount/input/images/Pancancer-lung'), ArgumentValue(name='feat_path_test', value='/mount/input/features/Pancancer'), ArgumentValue(name='clini_table', value='/mount/input/Pancancer_clini.xlsx'), ArgumentValue(name='target_label', value='Type'), ArgumentValue(name='base_dir', value='/mount/output/counterfactuals_lung_types'), ArgumentValue(name='manipulation_levels', value=[0.06]), ArgumentValue(name='pretrained_autoenc_name', value='pancancer_model'), ArgumentValue(name='pretrained_clf_name', value='lung')], mount=[Mount(source='images/Pancancer-lung', target='images/Pancancer-lung'), Mount(source='features/Pancancer', target='features/Pancancer'), Mount(source='Pancancer_clini.xlsx', target='Pancancer_clini.xlsx')])], note=None),\n",
       " ToolDefinition(name='musk_extract_features', repo=Repository(name='MUSK', url='https://github.com/lilab-stanford/MUSK', branch=None, commit='e1699c2', env=[EnvironmentVariable(name='HF_TOKEN', value='${env:HF_TOKEN}')]), requires='cuda', papers=['xiang2025musk'], category='pathology', description='Perform feature extraction on an input image using the vision part of MUSK.', arguments=[ArgumentDefinition(name='input_image', description='Path to the input image', type='str')], returns=[ArgumentDefinition(name='features', description='The feature vector extracted from the input image, as a list of floats', type='list')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='input_image', value='/mount/input/TUM/TUM-TCGA-ACRLPPQE.tif')], mount=[Mount(source='TUM-TCGA-ACRLPPQE.tif', target='TUM/TUM-TCGA-ACRLPPQE.tif')]), test_invocations=[ToolInvocation(name='kather100k_muc', arguments=[ArgumentValue(name='input_image', value='/mount/input/MUC/MUC-TCGA-ACCPKIPN.tif')], mount=[Mount(source='MUC-TCGA-ACCPKIPN.tif', target='MUC/MUC-TCGA-ACCPKIPN.tif')]), ToolInvocation(name='tcga_brca_patch_png', arguments=[ArgumentValue(name='input_image', value='/mount/input/TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.png')], mount=[Mount(source='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.png', target='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.png')]), ToolInvocation(name='tcga_brca_patch_jpg', arguments=[ArgumentValue(name='input_image', value='/mount/input/TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.jpg')], mount=[Mount(source='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.jpg', target='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.jpg')])], note=None),\n",
       " ToolDefinition(name='nnunet_preprocess', repo=Repository(name='nnUNet', url='https://github.com/MIC-DKFZ/nnUNet', branch=None, commit='58a3b12', env=[]), requires='cpu', papers=['isensee2020nnunet'], category='radiology', description='Preprocess a dataset using nnUNetv2. The dataset is in the old Medical Segmentation Decathlon (MSD) format and will need to be converted. Does not require GPU.', arguments=[ArgumentDefinition(name='dataset_path', description='The path to the dataset folder to train the model on (in MSD format, so contains dataset.json, imagesTr, imagesTs, labelsTr)', type='str')], returns=[ArgumentDefinition(name='dataset_json', description='The dataset config object (dataset.json) created by nnUNetv2, as the parsed json object', type='dict'), ArgumentDefinition(name='nnUNetPlans_json', description='The nnUNetv2 plan file (nnUNetPlans.json) created by nnUNetv2, as the parsed json object', type='dict')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='dataset_path', value='/mount/input/Task02_Heart')], mount=[Mount(source='msd/Task02_Heart', target='Task02_Heart')]), test_invocations=[ToolInvocation(name='prostate', arguments=[ArgumentValue(name='dataset_path', value='/mount/input/Task05_Prostate')], mount=[Mount(source='msd/Task05_Prostate', target='Task05_Prostate')]), ToolInvocation(name='spleen', arguments=[ArgumentValue(name='dataset_path', value='/mount/input/Task09_Spleen')], mount=[Mount(source='msd/Task09_Spleen', target='Task09_Spleen')]), ToolInvocation(name='hippocampus', arguments=[ArgumentValue(name='dataset_path', value='/mount/input/Task04_Hippocampus')], mount=[Mount(source='msd/Task04_Hippocampus', target='Task04_Hippocampus')])], note='Use the UNet model from DKFZ to train a medical segmentation model. More info here: https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/how_to_use_nnunet.md https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/dataset_format.md'),\n",
       " ToolDefinition(name='pathfinder_verify_biomarker', repo=Repository(name='PathFinder', url='https://github.com/LiangJunhao-THU/PathFinderCRC', branch=None, commit='093d77b', env=[]), requires='cpu', papers=['liang2023pathfinder'], category='pathology', description='Given WSI probability maps, a hypothesis of a potential biomarker, and clinical data, determine (1) whether the potential biomarker is significant for patient prognosis, and (2) whether the potential biomarker is independent among already known biomarkers.', arguments=[ArgumentDefinition(name='heatmaps', description='Path to the folder containing the numpy array (`*.npy`) files, which contains the heatmaps of the trained model (each heatmap is HxWxC where C is the number of classes)', type='str'), ArgumentDefinition(name='hypothesis', description='A python file, which contains a function `def hypothesis_score(prob_map_path: str) -> float` which expresses a mathematical model of a hypothesis of a potential biomarker.  For a particular patient (whose heatmap is given by `prob_map_path` as a npy file), the function returns a risk score.', type='str'), ArgumentDefinition(name='clini_table', description='Path to the CSV file containing the clinical data', type='str'), ArgumentDefinition(name='files_table', description='Path to the CSV file containing the mapping between patient IDs (in the PATIENT column) and heatmap filenames (in the FILENAME column)', type='str'), ArgumentDefinition(name='survival_time_column', description='The name of the column in the clinical data that contains the survival time', type='str'), ArgumentDefinition(name='event_column', description='The name of the column in the clinical data that contains the event (e.g. death, recurrence, etc.)', type='str'), ArgumentDefinition(name='known_biomarkers', description='A list of known biomarkers. These are column names in the clinical data.', type='list')], returns=[ArgumentDefinition(name='p_value', description='The p-value of the significance of the potential biomarker', type='float'), ArgumentDefinition(name='hazard_ratio', description='The hazard ratio for the biomarker', type='float')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='heatmaps', value='/mount/input/TCGA_CRC'), ArgumentValue(name='hypothesis', value='/mount/input/mus_fraction_score.py'), ArgumentValue(name='clini_table', value='/mount/input/TCGA_CRC_info.csv'), ArgumentValue(name='files_table', value='/mount/input/TCGA_CRC_files.csv'), ArgumentValue(name='survival_time_column', value='OS.time'), ArgumentValue(name='event_column', value='vital_status'), ArgumentValue(name='known_biomarkers', value=['MSI'])], mount=[Mount(source='TCGA_CRC', target='TCGA_CRC'), Mount(source='mus_fraction_score.py', target='mus_fraction_score.py'), Mount(source='TCGA_CRC_info.csv', target='TCGA_CRC_info.csv'), Mount(source='TCGA_CRC_files.csv', target='TCGA_CRC_files.csv')]), test_invocations=[ToolInvocation(name='crc_tum_fraction_score', arguments=[ArgumentValue(name='heatmaps', value='/mount/input/TCGA_CRC'), ArgumentValue(name='hypothesis', value='/mount/input/tum_fraction_score.py'), ArgumentValue(name='clini_table', value='/mount/input/TCGA_CRC_info.csv'), ArgumentValue(name='files_table', value='/mount/input/TCGA_CRC_files.csv'), ArgumentValue(name='survival_time_column', value='OS.time'), ArgumentValue(name='event_column', value='OS'), ArgumentValue(name='known_biomarkers', value=['MSI'])], mount=[Mount(source='TCGA_CRC', target='TCGA_CRC'), Mount(source='tum_fraction_score.py', target='tum_fraction_score.py'), Mount(source='TCGA_CRC_info.csv', target='TCGA_CRC_info.csv'), Mount(source='TCGA_CRC_files.csv', target='TCGA_CRC_files.csv')]), ToolInvocation(name='crc_str_fraction_score', arguments=[ArgumentValue(name='heatmaps', value='/mount/input/TCGA_CRC'), ArgumentValue(name='hypothesis', value='/mount/input/str_fraction_score.py'), ArgumentValue(name='clini_table', value='/mount/input/TCGA_CRC_info.csv'), ArgumentValue(name='files_table', value='/mount/input/TCGA_CRC_files.csv'), ArgumentValue(name='survival_time_column', value='OS.time'), ArgumentValue(name='event_column', value='OS'), ArgumentValue(name='known_biomarkers', value=['MSI'])], mount=[Mount(source='TCGA_CRC', target='TCGA_CRC'), Mount(source='str_fraction_score.py', target='str_fraction_score.py'), Mount(source='TCGA_CRC_info.csv', target='TCGA_CRC_info.csv'), Mount(source='TCGA_CRC_files.csv', target='TCGA_CRC_files.csv')]), ToolInvocation(name='cptac_str_fraction_score', arguments=[ArgumentValue(name='heatmaps', value='/mount/input/CPTAC_CRC'), ArgumentValue(name='hypothesis', value='/mount/input/str_fraction_score.py'), ArgumentValue(name='clini_table', value='/mount/input/CPTAC_CRC_info.csv'), ArgumentValue(name='files_table', value='/mount/input/CPTAC_CRC_files.csv'), ArgumentValue(name='survival_time_column', value='OS.time'), ArgumentValue(name='event_column', value='OS'), ArgumentValue(name='known_biomarkers', value=['MSI'])], mount=[Mount(source='CPTAC_CRC', target='CPTAC_CRC'), Mount(source='str_fraction_score.py', target='str_fraction_score.py'), Mount(source='CPTAC_CRC_info.csv', target='CPTAC_CRC_info.csv'), Mount(source='CPTAC_CRC_files.csv', target='CPTAC_CRC_files.csv')])], note=None),\n",
       " ToolDefinition(name='retfound_feature_vector', repo=Repository(name='RETFound', url='https://github.com/rmaphoh/RETFound_MAE', branch=None, commit='897d71c', env=[EnvironmentVariable(name='HF_TOKEN', value='${env:HF_TOKEN}')]), requires='cuda', papers=['zhou2023retfound'], category='misc', description='Extract the latent feature vector for the given retinal image using the RETFound pretrained RETFound_mae_natureCFP model.', arguments=[ArgumentDefinition(name='image_file', description='Path to the retinal image.', type='str')], returns=[ArgumentDefinition(name='feature_vector', description='The feature vector for the given retinal image, as a list of floats.', type='list')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='image_file', value='/mount/input/retinal_image.jpg')], mount=[Mount(source='cucumber.jpg', target='retinal_image.jpg')]), test_invocations=[ToolInvocation(name='jpg', arguments=[ArgumentValue(name='image_file', value='/mount/input/image1.jpg')], mount=[Mount(source='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.jpg', target='image1.jpg')]), ToolInvocation(name='png', arguments=[ArgumentValue(name='image_file', value='/mount/input/image2.png')], mount=[Mount(source='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.png', target='image2.png')]), ToolInvocation(name='cucumber_different_filename', arguments=[ArgumentValue(name='image_file', value='/mount/input/cucumber.jpg')], mount=[Mount(source='cucumber.jpg', target='cucumber.jpg')])], note=None),\n",
       " ToolDefinition(name='stamp_extract_features', repo=Repository(name='STAMP', url='https://github.com/KatherLab/STAMP', branch=None, commit='97522aa', env=[]), requires='cuda', papers=['elnahhas2024stamp'], category='pathology', description='Perform feature extraction using CTransPath with STAMP on a set of whole slide images, saving the resulting features to the specified output directory.', arguments=[ArgumentDefinition(name='output_dir', description='Path to the output folder where the features will be saved', type='str'), ArgumentDefinition(name='slide_dir', description='Path to the input folder containing the whole slide images', type='str')], returns=[ArgumentDefinition(name='num_processed_slides', description='The number of slides that were processed', type='int')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='output_dir', value='/mount/output/TCGA-BRCA-features'), ArgumentValue(name='slide_dir', value='/mount/input/TCGA-BRCA-SLIDES')], mount=[Mount(source='brca', target='TCGA-BRCA-SLIDES')]), test_invocations=[ToolInvocation(name='crc', arguments=[ArgumentValue(name='output_dir', value='/mount/output/TCGA-CRC-features'), ArgumentValue(name='slide_dir', value='/mount/input/TCGA-CRC-SLIDES')], mount=[Mount(source='crc', target='TCGA-CRC-SLIDES')]), ToolInvocation(name='crc_single', arguments=[ArgumentValue(name='output_dir', value='/mount/output/TCGA-CRC-features'), ArgumentValue(name='slide_dir', value='/mount/input/TCGA-CRC-SLIDES')], mount=[Mount(source='crc/TCGA-4N-A93T-01Z-00-DX1.82E240B1-22C3-46E3-891F-0DCE35C43F8B.svs', target='TCGA-CRC-SLIDES/TCGA-4N-A93T-01Z-00-DX1.82E240B1-22C3-46E3-891F-0DCE35C43F8B.svs')]), ToolInvocation(name='brca_single', arguments=[ArgumentValue(name='output_dir', value='/mount/output/TCGA-BRCA-features'), ArgumentValue(name='slide_dir', value='/mount/input/TCGA-BRCA-SLIDES')], mount=[Mount(source='brca/TCGA-BH-A0BZ-01Z-00-DX1.45EB3E93-A871-49C6-9EAE-90D98AE01913.svs', target='TCGA-BRCA-SLIDES/TCGA-BH-A0BZ-01Z-00-DX1.45EB3E93-A871-49C6-9EAE-90D98AE01913.svs')])], note=None),\n",
       " ToolDefinition(name='stamp_train_classification_model', repo=Repository(name='STAMP', url='https://github.com/KatherLab/STAMP', branch=None, commit='97522aa', env=[]), requires='cuda', papers=['elnahhas2024stamp'], category='pathology', description='Train a model for biomarker classification. You will be supplied with the path to the folder containing the whole slide images, alongside a path to a CSV file containing the training labels. Use ctranspath for feature extraction.', arguments=[ArgumentDefinition(name='slide_dir', description='Path to the folder containing the whole slide images', type='str'), ArgumentDefinition(name='clini_table', description='Path to the CSV file containing the clinical data', type='str'), ArgumentDefinition(name='slide_table', description='Path to the CSV file containing the slide metadata', type='str'), ArgumentDefinition(name='target_column', description='The name of the column in the clinical data that contains the target labels', type='str'), ArgumentDefinition(name='trained_model_path', description='Path to the *.ckpt file where the trained model should be saved by this function', type='str')], returns=[ArgumentDefinition(name='num_params', description='The number of parameters in the trained model', type='int')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='slide_dir', value='/mount/input/TCGA-BRCA-SLIDES'), ArgumentValue(name='clini_table', value='/mount/input/TCGA-BRCA-DX_CLINI.xlsx'), ArgumentValue(name='slide_table', value='/mount/input/TCGA-BRCA-DX_SLIDE.csv'), ArgumentValue(name='target_column', value='TP53_driver'), ArgumentValue(name='trained_model_path', value='/mount/output/STAMP-BRCA-TP53-model.ckpt')], mount=[Mount(source='brca', target='TCGA-BRCA-SLIDES'), Mount(source='TCGA-BRCA-DX_CLINI.xlsx', target='TCGA-BRCA-DX_CLINI.xlsx'), Mount(source='TCGA-BRCA-DX_SLIDE.csv', target='TCGA-BRCA-DX_SLIDE.csv')]), test_invocations=[ToolInvocation(name='crc_msi', arguments=[ArgumentValue(name='slide_dir', value='/mount/input/TCGA-CRC-SLIDES'), ArgumentValue(name='clini_table', value='/mount/input/TCGA-CRC-DX_CLINI.xlsx'), ArgumentValue(name='slide_table', value='/mount/input/TCGA-CRC-DX_SLIDE.csv'), ArgumentValue(name='target_column', value='isMSIH'), ArgumentValue(name='trained_model_path', value='/mount/output/STAMP-CRC-MSI-model.ckpt')], mount=[Mount(source='crc', target='TCGA-CRC-SLIDES'), Mount(source='TCGA-CRC-DX_CLINI.xlsx', target='TCGA-CRC-DX_CLINI.xlsx'), Mount(source='TCGA-CRC-DX_SLIDE.csv', target='TCGA-CRC-DX_SLIDE.csv')]), ToolInvocation(name='crc_braf', arguments=[ArgumentValue(name='slide_dir', value='/mount/input/TCGA-CRC-SLIDES'), ArgumentValue(name='clini_table', value='/mount/input/TCGA-CRC-DX_CLINI.xlsx'), ArgumentValue(name='slide_table', value='/mount/input/TCGA-CRC-DX_SLIDE.csv'), ArgumentValue(name='target_column', value='BRAF'), ArgumentValue(name='trained_model_path', value='/mount/output/STAMP-CRC-BRAF-model.ckpt')], mount=[Mount(source='crc', target='TCGA-CRC-SLIDES'), Mount(source='TCGA-CRC-DX_CLINI.xlsx', target='TCGA-CRC-DX_CLINI.xlsx'), Mount(source='TCGA-CRC-DX_SLIDE.csv', target='TCGA-CRC-DX_SLIDE.csv')]), ToolInvocation(name='crc_kras', arguments=[ArgumentValue(name='slide_dir', value='/mount/input/TCGA-CRC-SLIDES'), ArgumentValue(name='clini_table', value='/mount/input/TCGA-CRC-DX_CLINI.xlsx'), ArgumentValue(name='slide_table', value='/mount/input/TCGA-CRC-DX_SLIDE.csv'), ArgumentValue(name='target_column', value='KRAS'), ArgumentValue(name='trained_model_path', value='/mount/output/STAMP-CRC-KRAS-model.ckpt')], mount=[Mount(source='crc', target='TCGA-CRC-SLIDES'), Mount(source='TCGA-CRC-DX_CLINI.xlsx', target='TCGA-CRC-DX_CLINI.xlsx'), Mount(source='TCGA-CRC-DX_SLIDE.csv', target='TCGA-CRC-DX_SLIDE.csv')])], note='Here, the agent must realize that it needs to perform feature extraction before training the model.'),\n",
       " ToolDefinition(name='tabpfn_predict', repo=Repository(name='TabPFN', url='https://github.com/PriorLabs/TabPFN', branch=None, commit='e8744e4', env=[]), requires='cpu', papers=['hollmann2025tabpfn'], category='misc', description='Train a predictor using TabPFN on a tabular dataset. Evaluate the predictor on the test set. Train on CPU.', arguments=[ArgumentDefinition(name='train_csv', description='Path to the CSV file containing the training data', type='str'), ArgumentDefinition(name='test_csv', description='Path to the CSV file containing the test data', type='str'), ArgumentDefinition(name='feature_columns', description='The names of the columns to use as features', type='list'), ArgumentDefinition(name='target_column', description='The name of the column to predict', type='str')], returns=[ArgumentDefinition(name='roc_auc', description='The ROC AUC score of the predictor on the test set', type='float'), ArgumentDefinition(name='accuracy', description='The accuracy of the predictor on the test set', type='float'), ArgumentDefinition(name='probs', description='The probabilities of the predictor on the test set, as a list of floats (one per sample in the test set)', type='list')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='train_csv', value='/mount/input/breast_cancer_train.csv'), ArgumentValue(name='test_csv', value='/mount/input/breast_cancer_test.csv'), ArgumentValue(name='feature_columns', value=['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension', 'radius error', 'texture error', 'perimeter error', 'area error', 'smoothness error', 'compactness error', 'concavity error', 'concave points error', 'symmetry error', 'fractal dimension error', 'worst radius', 'worst texture', 'worst perimeter', 'worst area', 'worst smoothness', 'worst compactness', 'worst concavity', 'worst concave points', 'worst symmetry', 'worst fractal dimension']), ArgumentValue(name='target_column', value='target')], mount=[Mount(source='breast_cancer_train.csv', target='breast_cancer_train.csv'), Mount(source='breast_cancer_test.csv', target='breast_cancer_test.csv')]), test_invocations=[ToolInvocation(name='diabetes', arguments=[ArgumentValue(name='train_csv', value='/mount/input/diabetes_train.csv'), ArgumentValue(name='test_csv', value='/mount/input/diabetes_test.csv'), ArgumentValue(name='feature_columns', value=['preg', 'plas', 'pres', 'skin', 'insu', 'mass', 'pedi', 'age']), ArgumentValue(name='target_column', value='class')], mount=[Mount(source='diabetes_train.csv', target='diabetes_train.csv'), Mount(source='diabetes_test.csv', target='diabetes_test.csv')]), ToolInvocation(name='heart_disease', arguments=[ArgumentValue(name='train_csv', value='/mount/input/heart_disease_train.csv'), ArgumentValue(name='test_csv', value='/mount/input/heart_disease_test.csv'), ArgumentValue(name='feature_columns', value=['age', 'sex', 'chest', 'resting_blood_pressure', 'serum_cholestoral', 'fasting_blood_sugar', 'resting_electrocardiographic_results', 'maximum_heart_rate_achieved', 'exercise_induced_angina', 'oldpeak', 'slope', 'number_of_major_vessels', 'thal']), ArgumentValue(name='target_column', value='class')], mount=[Mount(source='heart_disease_train.csv', target='heart_disease_train.csv'), Mount(source='heart_disease_test.csv', target='heart_disease_test.csv')]), ToolInvocation(name='parkinsons', arguments=[ArgumentValue(name='train_csv', value='/mount/input/parkinsons_train.csv'), ArgumentValue(name='test_csv', value='/mount/input/parkinsons_test.csv'), ArgumentValue(name='feature_columns', value=['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22']), ArgumentValue(name='target_column', value='Class')], mount=[Mount(source='parkinsons_train.csv', target='parkinsons_train.csv'), Mount(source='parkinsons_test.csv', target='parkinsons_test.csv')])], note=None),\n",
       " ToolDefinition(name='textgrad_medical_qa_optimize', repo=Repository(name='textgrad', url='https://github.com/zou-group/textgrad', branch=None, commit='bf5b0c5', env=[EnvironmentVariable(name='OPENAI_API_KEY', value='${env:OPENAI_API_KEY}')]), requires='cpu', papers=['yuksekgonul2024textgrad'], category='llms', description='Optimize answers to multiple-choice medical questions using TextGrad. Each question is improved at test-time through textual gradients, guided by an objective (e.g. \"Make the answer concise and accurate\"). \\n', arguments=[ArgumentDefinition(name='csv_path', description=\"Path to a CSV file containing columns: 'index', 'question', 'objective'\", type='str'), ArgumentDefinition(name='backward_engine', description=\"The model used to compute textual gradients (e.g., 'gpt-4o')\", type='str'), ArgumentDefinition(name='forward_engine', description=\"The model used to generate initial zero-shot answers (e.g., 'gpt-3.5-turbo')\", type='str'), ArgumentDefinition(name='starting_system_prompt', description='System prompt to guide the LLM behavior', type='str'), ArgumentDefinition(name='optimizer_constraint', description='Constraint string that specifies the required answer format', type='str')], returns=[ArgumentDefinition(name='optimized_answers', description=\"A list of optimized answers, each ending with 'Answer: $LETTER'\", type='list')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='csv_path', value='/mount/input/sample_0.csv'), ArgumentValue(name='backward_engine', value='gpt-4o'), ArgumentValue(name='forward_engine', value='gpt-4o'), ArgumentValue(name='starting_system_prompt', value='You are ChatGPT, a large language model trained by OpenAI, based on the GPT-4 architecture.\\nKnowledge cutoff: 2023-12\\nCurrent date: 2024-04-01\\n'), ArgumentValue(name='optimizer_constraint', value=\"You must end your answer with a separate line like: 'Answer: A', 'Answer: B', 'Answer: C', or 'Answer: D'. \\nDo NOT include any additional explanation or diagnosis after 'Answer: $LETTER'.\\n\")], mount=[Mount(source='sample_0.csv', target='sample_0.csv')]), test_invocations=[ToolInvocation(name='sample_0', arguments=[ArgumentValue(name='csv_path', value='/mount/input/sample_0.csv'), ArgumentValue(name='backward_engine', value='gpt-4o'), ArgumentValue(name='forward_engine', value='gpt-4'), ArgumentValue(name='starting_system_prompt', value='You are ChatGPT, a large language model trained by OpenAI, based on the GPT-4 architecture.\\nKnowledge cutoff: 2023-12\\nCurrent date: 2024-04-01\\n'), ArgumentValue(name='optimizer_constraint', value=\"You must end your answer with a separate line like: 'Answer: A', 'Answer: B', 'Answer: C', or 'Answer: D'. \\nDo NOT include any additional explanation or diagnosis after 'Answer: $LETTER'.\\n\")], mount=[Mount(source='sample_0.csv', target='sample_0.csv')]), ToolInvocation(name='sample_1', arguments=[ArgumentValue(name='csv_path', value='/mount/input/sample_1.csv'), ArgumentValue(name='backward_engine', value='gpt-4o'), ArgumentValue(name='forward_engine', value='gpt-4'), ArgumentValue(name='starting_system_prompt', value='You are ChatGPT, a large language model trained by OpenAI, based on the GPT-4 architecture.\\nKnowledge cutoff: 2023-12\\nCurrent date: 2024-04-01\\n'), ArgumentValue(name='optimizer_constraint', value=\"You must end your answer with a separate line like: 'Answer: A', 'Answer: B', 'Answer: C', or 'Answer: D'. \\nDo NOT include any additional explanation or diagnosis after 'Answer: $LETTER'.\\n\")], mount=[Mount(source='sample_1.csv', target='sample_1.csv')]), ToolInvocation(name='sample_2', arguments=[ArgumentValue(name='csv_path', value='/mount/input/sample_2.csv'), ArgumentValue(name='backward_engine', value='gpt-4o'), ArgumentValue(name='forward_engine', value='gpt-4'), ArgumentValue(name='starting_system_prompt', value='You are ChatGPT, a large language model trained by OpenAI, based on the GPT-4 architecture.\\nKnowledge cutoff: 2023-12\\nCurrent date: 2024-04-01\\n'), ArgumentValue(name='optimizer_constraint', value=\"You must end your answer with a separate line like: 'Answer: A', 'Answer: B', 'Answer: C', or 'Answer: D'. \\nDo NOT include any additional explanation or diagnosis after 'Answer: $LETTER'.\\n\")], mount=[Mount(source='sample_2.csv', target='sample_2.csv')])], note=None),\n",
       " ToolDefinition(name='tiatoolbox_wsi_dimensions', repo=Repository(name='tiatoolbox', url='https://github.com/TissueImageAnalytics/tiatoolbox', branch=None, commit='7ba7394', env=[]), requires='cpu', papers=['pocock2022tiatoolbox'], category='pathology', description='Determine the pixel dimensions for every whole slide image (WSI) in `input_dir` using TIAToolbox.', arguments=[ArgumentDefinition(name='input_dir', description='Path to the folder that contains the WSIs', type='str'), ArgumentDefinition(name='include_pyramid', description='Whether to include every pyramid level instead of only the baseline dimensions', type='bool')], returns=[ArgumentDefinition(name='dimensions', description='Dimensions of the WSI (optionally with of without full pyramid values) as a dict of  {slide_filename: {\"baseline\": [width, height], \"levels\": [[width, height], ...]}}, where  `baseline` is the dimensions of the WSI at the highest resolution and `levels` is a list of  dimensions for each pyramid level. If `include_pyramid` is `False`, only the `baseline`  dimensions are included.', type='dict')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='input_dir', value='/mount/input/wsis'), ArgumentValue(name='include_pyramid', value=True)], mount=[Mount(source='wsis', target='wsis')]), test_invocations=[ToolInvocation(name='single_wsi_baseline', arguments=[ArgumentValue(name='input_dir', value='/mount/input/wsis'), ArgumentValue(name='include_pyramid', value=False)], mount=[Mount(source='wsis/TCGA-DT-5265-01Z-00-DX1.563f09af-8bbe-45cd-9c6d-85a96255e67f.svs', target='wsis/TCGA-DT-5265-01Z-00-DX1.563f09af-8bbe-45cd-9c6d-85a96255e67f.svs')]), ToolInvocation(name='single_wsi_full_pyramid', arguments=[ArgumentValue(name='input_dir', value='/mount/input/wsis'), ArgumentValue(name='include_pyramid', value=True)], mount=[Mount(source='wsis/TCGA-DT-5265-01Z-00-DX1.563f09af-8bbe-45cd-9c6d-85a96255e67f.svs', target='wsis/TCGA-DT-5265-01Z-00-DX1.563f09af-8bbe-45cd-9c6d-85a96255e67f.svs')]), ToolInvocation(name='two_wsi_full_pyramid', arguments=[ArgumentValue(name='input_dir', value='/mount/input/wsis'), ArgumentValue(name='include_pyramid', value=True)], mount=[Mount(source='wsis/TCGA-AG-A011-01Z-00-DX1.155A4093-5EC6-4D38-8CE1-24C045DF0CD8.svs', target='wsis/TCGA-AG-A011-01Z-00-DX1.155A4093-5EC6-4D38-8CE1-24C045DF0CD8.svs'), Mount(source='wsis/TCGA-EI-6881-01Z-00-DX1.5cfa2929-4374-4166-b110-39ab7d3de7cd.svs', target='wsis/TCGA-EI-6881-01Z-00-DX1.5cfa2929-4374-4166-b110-39ab7d3de7cd.svs')]), ToolInvocation(name='full_dir_baseline', arguments=[ArgumentValue(name='input_dir', value='/mount/input/wsis'), ArgumentValue(name='include_pyramid', value=False)], mount=[Mount(source='wsis', target='wsis')])], note=None),\n",
       " ToolDefinition(name='tiatoolbox_wsi_thumbnailer', repo=Repository(name='tiatoolbox', url='https://github.com/TissueImageAnalytics/tiatoolbox', branch=None, commit='7ba7394', env=[]), requires='cpu', papers=['pocock2022tiatoolbox'], category='pathology', description='Generate a PNG thumbnail for every whole-slide image (WSI) in `input_dir` using TIAToolbox and save them to `output_dir` with the suffix _thumbnail.png.', arguments=[ArgumentDefinition(name='input_dir', description='Path to the folder that contains the WSIs', type='str'), ArgumentDefinition(name='output_dir', description='Path to the folder where thumbnails are written', type='str'), ArgumentDefinition(name='resolution', description='Requested magnification / physical resolution', type='float'), ArgumentDefinition(name='units', description='Units for resolution (\"power\", \"mpp\", \"level\", \"baseline\")', type='str')], returns=[ArgumentDefinition(name='num_thumbnails', description='Number of thumbnails created', type='int')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='input_dir', value='/mount/input/wsis'), ArgumentValue(name='output_dir', value='/mount/output/wsis_thumbs'), ArgumentValue(name='resolution', value=1.25), ArgumentValue(name='units', value='power')], mount=[Mount(source='wsis', target='wsis')]), test_invocations=[ToolInvocation(name='single_wsi_low_power', arguments=[ArgumentValue(name='input_dir', value='/mount/input/wsis'), ArgumentValue(name='output_dir', value='/mount/output/wsis_thumbs'), ArgumentValue(name='resolution', value=0.625), ArgumentValue(name='units', value='power')], mount=[Mount(source='wsis/TCGA-DT-5265-01Z-00-DX1.563f09af-8bbe-45cd-9c6d-85a96255e67f.svs', target='wsis/TCGA-DT-5265-01Z-00-DX1.563f09af-8bbe-45cd-9c6d-85a96255e67f.svs')]), ToolInvocation(name='two_wsi_at_2mpp', arguments=[ArgumentValue(name='input_dir', value='/mount/input/wsis'), ArgumentValue(name='output_dir', value='/mount/output/wsis_thumbs'), ArgumentValue(name='resolution', value=2.0), ArgumentValue(name='units', value='mpp')], mount=[Mount(source='wsis/TCGA-AG-A011-01Z-00-DX1.155A4093-5EC6-4D38-8CE1-24C045DF0CD8.svs', target='wsis/TCGA-AG-A011-01Z-00-DX1.155A4093-5EC6-4D38-8CE1-24C045DF0CD8.svs'), Mount(source='wsis/TCGA-EI-6881-01Z-00-DX1.5cfa2929-4374-4166-b110-39ab7d3de7cd.svs', target='wsis/TCGA-EI-6881-01Z-00-DX1.5cfa2929-4374-4166-b110-39ab7d3de7cd.svs')]), ToolInvocation(name='full_dir_1p25_power', arguments=[ArgumentValue(name='input_dir', value='/mount/input/wsis'), ArgumentValue(name='output_dir', value='/mount/output/wsis_thumbs'), ArgumentValue(name='resolution', value=1.25), ArgumentValue(name='units', value='power')], mount=[Mount(source='wsis', target='wsis')])], note=None),\n",
       " ToolDefinition(name='totalsegmentator_segment_liver', repo=Repository(name='TotalSegmentator', url='https://github.com/wasserth/TotalSegmentator', branch=None, commit='5b1a4f0', env=[]), requires='cuda', papers=['wasserthal2023totalsegmentator'], category='radiology', description='Segment the liver vessels from the input CT scan(s) and save the result as a .nii.gz file. In the output segmentation, liver vessel voxels must have the value 1, and all other voxels must be 0.', arguments=[ArgumentDefinition(name='ct_input_path', description='Path to the input image in .nii.gz format', type='str'), ArgumentDefinition(name='segmentation_mask_output_path', description='Path to the output file (liver vessels segmentation mask) in .nii.gz format', type='str')], returns=[], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='ct_input_path', value='/mount/input/CRLM-CT-1040_0000.nii.gz'), ArgumentValue(name='segmentation_mask_output_path', value='/mount/output/CRLM-CT-1040_0000_seg_mask.nii.gz')], mount=[Mount(source='CRLM-CT-1040_0000.nii.gz', target='CRLM-CT-1040_0000.nii.gz')]), test_invocations=[ToolInvocation(name='CRLM-CT-1085', arguments=[ArgumentValue(name='ct_input_path', value='/mount/input/CRLM-CT-1085_0000.nii.gz'), ArgumentValue(name='segmentation_mask_output_path', value='/mount/output/CRLM-CT-1085_0000_seg_mask.nii.gz')], mount=[Mount(source='CRLM-CT-1085_0000.nii.gz', target='CRLM-CT-1085_0000.nii.gz')]), ToolInvocation(name='CRLM-CT-1161', arguments=[ArgumentValue(name='ct_input_path', value='/mount/input/CRLM-CT-1161_0000.nii.gz'), ArgumentValue(name='segmentation_mask_output_path', value='/mount/output/CRLM-CT-1161_0000_seg_mask.nii.gz')], mount=[Mount(source='CRLM-CT-1161_0000.nii.gz', target='CRLM-CT-1161_0000.nii.gz')]), ToolInvocation(name='CRLM-CT-1094', arguments=[ArgumentValue(name='ct_input_path', value='/mount/input/CRLM-CT-1094_0000.nii.gz'), ArgumentValue(name='segmentation_mask_output_path', value='/mount/output/CRLM-CT-1094_0000_seg_mask.nii.gz')], mount=[Mount(source='CRLM-CT-1094_0000.nii.gz', target='CRLM-CT-1094_0000.nii.gz')])], note=None),\n",
       " ToolDefinition(name='uni_extract_features', repo=Repository(name='UNI', url='https://github.com/mahmoodlab/UNI', branch=None, commit='42715ef', env=[EnvironmentVariable(name='HF_TOKEN', value='${env:HF_TOKEN}')]), requires='cuda', papers=['chen2024uni'], category='pathology', description='Perform feature extraction on an input image using the \"UNI\" model.', arguments=[ArgumentDefinition(name='input_image', description='Path to the input image', type='str')], returns=[ArgumentDefinition(name='features', description='The feature vector extracted from the input image, as a list of floats', type='list')], example=ExampleInvocation(name='example', arguments=[ArgumentValue(name='input_image', value='/mount/input/TUM/TUM-TCGA-ACRLPPQE.tif')], mount=[Mount(source='TUM-TCGA-ACRLPPQE.tif', target='TUM/TUM-TCGA-ACRLPPQE.tif')]), test_invocations=[ToolInvocation(name='kather100k_muc', arguments=[ArgumentValue(name='input_image', value='/mount/input/MUC/MUC-TCGA-ACCPKIPN.tif')], mount=[Mount(source='MUC-TCGA-ACCPKIPN.tif', target='MUC/MUC-TCGA-ACCPKIPN.tif')]), ToolInvocation(name='tcga_brca_patch_png', arguments=[ArgumentValue(name='input_image', value='/mount/input/TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.png')], mount=[Mount(source='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.png', target='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.png')]), ToolInvocation(name='tcga_brca_patch_jpg', arguments=[ArgumentValue(name='input_image', value='/mount/input/TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.jpg')], mount=[Mount(source='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.jpg', target='TCGA-BRCA_patch_TCGA-BH-A0DE-01Z-00-DX1.64A0340A-8146-48E8-AAF7-4035988B9152.jpg')])], note=None)]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from toolarena import load_tasks\n",
    "\n",
    "ds = load_tasks(source=\"huggingface\")\n",
    "ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
